expression-language.md

```diff
@@ -0,0 +1,375 @@
+"OPTIONAL" in this document are to be interpreted as described in [RFC2119][rfc2119].
+
+### 1.2. Relation to the Subscriptions API
+
+The CESQL can be used as a [filter dialect][subscriptions-filter-dialect] to filter on the input values.
+
+<!-- TODO -->
+
+## 2. Language syntax
+
+The grammar of the language is defined using the EBNF Notation from [W3C XML specification][ebnf-xml-spec].
+
+Although in the EBNFs keywords are defined using uppercase characters, they are case-insensitive. For example:
+
+```
+int(hop) < int(ttl) and int(hop) < 1000
+```
+
+Has the same syntactical meaning of:
+
+```
+INT(hop) < INT(ttl) AND INT(hop) < 1000
+```
+
+### 2.1 Expression
+
+The root of the expression is the `expression` rule:
+
+```ebnf
+expression ::= value-identifier | literal | unary-operation | binary-operation | function-invocation | ( "(" expression ")" )
+```
+
+Nested expressions MUST be correctly parenthesized.
+
+### 2.2. Value identifiers and literals
+
+Value identifiers in CESQL have the same restrictions of the [Attribute Naming
+Convention][ce-attribute-naming-convention] from the CloudEvents spec. A value identifier MUST NOT be greater than 20
+characters in length.
+
+```ebnf
+lowercase-char ::= [a-z]
+value-identifier ::= lowercase-char ( lowercase-char | digit )*
+```
+
+CESQL defines 3 different literal kinds: integer numbers, `true` or `false` booleans and `''` or `""` delimited strings.
+
+```ebnf
+digit ::= [0-9]
+number-literal ::= digit+
+
+boolean-literal ::= "true" | "false"
+
+string-literal ::= ( "'" ( [^'] | "\'" )* "'" ) | ( '"' ( [^"] | '\"' )* '"')
+
+literal ::= number-literal | boolean-literal | string-literal
+```
+
+Because string literals can be either `''` or `""` delimited, in the former case, the `"` has to be escaped, while in
+the latter the `"` has to be escaped.
+
+### 2.3. Operators
+
+CESQL defines boolean unary and binary operators, arithmetic unary and binary operators and the `LIKE`, `IN`, `EXISTS`
+operators.
+
+```ebnf
+not-operator = "NOT"
+unary-logical-operator = not-operator
+binary-logic-operator = "AND" | "OR" | "XOR"
+
+unary-numeric-operator = "-"
+binary-comparison-operator = "=" | "!=" | "<>" | ">=" | "<=" | "<" | ">"
+binary-numeric-arithmetic-operator = "+" | "-" | "*" | "/" | "%"
+
+like-operator = "LIKE"
+exists-operator = "EXISTS"
+in-operator = "IN"
+
+unary-operation ::= (unary-numeric-operator | unary-logical-operator) expression
+
+binary-operation ::= expression (binary-comparison-operator | binary-numeric-arithmetic-operator | binary-logic-operator) expression
+
+like-operation ::= expression not-operator? like-operator string-literal
+
+exists-operation ::= exists-operator value-identifier
+
+set-expression ::= "(" expression ("," expression)* ")"
+in-operation ::= expression not-operator? in-operator set-expression
+```
+
+Note: the `set-expression` supports only compile-time set definition, hence the language doesn't include a _set_ type
```

Comment:
I don't think that statement is necessary, because the language only allows for expressions. You can't manipulate anything at runtime.

Reply:
Well, if we decide to have a set type, then an expression could return a set. I'm just underlining that, given this type doesn't exist, sets are only a compile time thing.

Reply:
can you reorder the EBNF rules so that the child nodes appear after the parent nodes? For instance `unary-operation` and `binary-operation` should appear first.

Reply:
since the language allows the creation of set then there must be a type set. You can't have one a without the other. Otherwise is the type of the value constructed by `set-expression` ?

Reply:
> since the language allows the creation of set then there must be a type set. You can't have one a without the other. Otherwise is the type of the value constructed by set-expression ?

The language doesn't allow for the definition of a set, that's exactly what i was talking about. Syntactically and semantically speaking, the set-expression is just part of the _in_ operator definition. I can just take the set-expression ebnf and collapse it in side the in-operation and things won't change.

Reply:
> The language doesn't allow for the definition of a set,

isn't what `set-expression` is doing, creating a set?  

>  I can just take the set-expression ebnf and collapse it in side the in-operation and things won't change

Inlining does not change the fact the comma operator in `(" expression ("," expression)* ")"` constructs a set (as you call it).

Reply:
> isn't what set-expression is doing, creating a set?

Yes, but you can't use it anywhere else than in the _IN_ operator. For me this is a detail of the definition of _IN_, not an actual type to be included in the type system. It's just a list of parameters for the _IN_ operator to compute.

Reply:
> Yes, but you can't use it anywhere else than in the IN operator

It does not matter how often it is used. This is perfectly normal for a programming language to define more types than the input it can process (eg. SQL, XPath, XQuery, etc..). 

> It's just a list of parameters for the IN operator to compute.

Variadic parameter lists are typically syntactic sugar for constructing a list/set via the function call syntax.

Reply:
So how about I just remove this sentence and that's it? Or do you think we should have a proper definition in the _Type system_ paragraph when we talk about the supported types?

Reply:
+1 to remove this sentence and to have a proper definition in the type system paragraph.

Reply:
@lionelvillard look at this commit https://github.com/cloudevents/spec/pull/771/commits/b1d75246ceff146fc83bc21497a18affaca1d751

I've defined set as a collection of strings, in order to restrict it to exactly what `IN` is capable of